src/uid2Sdk.ts

```diff
@@ -1,230 +1,354 @@
-    return expiry <= now;
+import { Uid2ApiClient } from "./uid2ApiClient";
+import {
+  EventType,
+  Uid2CallbackHandler,
+  Uid2CallbackManager,
+} from "./uid2CallbackManager";
+import { UID2CookieManager } from "./uid2CookieManager";
+import { Uid2Identity } from "./Uid2Identity";
+import { IdentityStatus, notifyInitCallback } from "./Uid2InitCallbacks";
+import { isUID2OptionsOrThrow, Uid2Options } from "./Uid2Options";
+import { UID2PromiseHandler } from "./uid2PromiseHandler";
+
+function hasExpired(expiry: number, now = Date.now()) {
+  return expiry <= now;
 }
 
 let postUid2CreateCallback: null | (() => void) = null;
 
 export class UID2 {
-    
-    static get VERSION() {
-        return "3.0.0";
-    }
-    static get COOKIE_NAME() {
-        return "__uid_2";
+  static get VERSION() {
+    return "3.0.0";
+  }
+  static get COOKIE_NAME() {
+    return "__uid_2";
+  }
+  static get DEFAULT_REFRESH_RETRY_PERIOD_MS() {
+    return 2000;
+  }
+  static IdentityStatus = IdentityStatus;
+  static EventType = EventType;
+
+  static setupGoogleTag() {
+    UID2.setupGoogleSecureSignals();
+  }
+
+  static setupGoogleSecureSignals() {
+    if (window.__uid2SecureSignalProvider)
+      window.__uid2SecureSignalProvider.registerSecureSignalProvider();
+  }
+
+  // Push functions to this array to receive event notifications
+  public callbacks: Uid2CallbackHandler[] = [];
+
+  // Dependencies initialised on construction
+  private _tokenPromiseHandler: UID2PromiseHandler;
+  private _callbackManager: Uid2CallbackManager;
+
+  // Dependencies initialised on call to init due to requirement for options
+  private _cookieManager: UID2CookieManager | undefined;
+  private _apiClient: Uid2ApiClient | undefined;
+
+  // State
+  private _opts: Uid2Options = {};
+  private _identity: Uid2Identity | null | undefined;
+  private _initComplete = false;
+
+  constructor(
+    existingCallbacks: Uid2CallbackHandler[] | undefined = undefined
+  ) {
+    if (existingCallbacks) this.callbacks = existingCallbacks;
+
+    this._tokenPromiseHandler = new UID2PromiseHandler(this);
+    this._callbackManager = new Uid2CallbackManager(this, () =>
+      this.getIdentity()
+    );
+    const runCallbacks = () => {
+      this._callbackManager.runCallbacks(EventType.SdkLoaded, {});
+    };
+    if (window.__uid2 instanceof UID2) {
+      runCallbacks();
+    } else {
+      // Need to defer running callbacks until this is assigned to the window global
+      postUid2CreateCallback = runCallbacks;
     }
-    static get DEFAULT_REFRESH_RETRY_PERIOD_MS() {
-        return 5000;
+  }
+
+  public init(opts: Uid2Options) {
+    this.initInternal(opts);
+  }
+  public getAdvertisingToken() {
+    return this.getIdentity()?.advertising_token ?? undefined;
+  }
+  public setIdentity(identity: Uid2Identity) {
+    const validatedIdentity = this.validateAndSetIdentity(identity);
+    if (validatedIdentity) {
+      this.triggerRefreshOrSetTimer(validatedIdentity);
```

Comment:
There might already be a refresh timer running when this is called. Rather than cancelling it here, it's worth thinking about whether the `setRefreshTimer` method should cancel any currently active timer - it doesn't make sense to have 2 of them active.

Reply:
I think there's another scenario that needs to be addressed.

1. SDK is initialized and loads an old identity from the cookie. It needs to be refreshed so it sends a refresh request.
2. The publisher sets a new identity which doesn't need to be refreshed. The timer starts to check again in 5s.
3. The refresh request from step 1 completes and the refreshed identity from the API call (based on the cookie) is set, so the identity the publisher set in step 2 is lost.
4. When the timer triggers, it will see the refreshed identity, not the new one provided in the call to setIdentity.

Reply:
In this case, the refreshed token from step 1 is the most recent identity, shouldn't we use it? The case that is problematic would be if the identity from the cookie and from the publisher is a different user? but this shouldn't happen if publisher use SDK correctly?

Reply:
It could happen, if someone switches user on a site. Or if people share a computer. I think the impact of this could be quite bad - if someone thinks they changed the user that's signed in, but they get targeted advertising for the previous user.

Remember we can't control how the publisher uses the SDK. If we give them a way to switch identity, but sometimes it puts them back to the old identity, that's a bug.

Think about a site which asks you "Is this you?" when you visit (I can't think of examples now but I see that a lot), and if you choose "no" it lets you pick from several users who sign in there sometimes. We don't want to undo the setIdentity when the refresh finishes.